utils/swift_build_support/swift_build_support/toolchain.py

```diff
@@ -0,0 +1,131 @@
+        # For details on `sysctl`, see:
+        # http://www.freebsd.org/cgi/man.cgi?sysctl(8)
+        return int(subprocess.check_output(
+            ['sysctl', '-n', 'kern.osreldate']).rstrip())
+    except subprocess.CalledProcessError:
+        return None
+
+
+def _first_common_toolchain(tools, suffixes=None):
+    """
+    Return a Toolchain of resolved paths where each path has
+    the same suffix.
+
+    If there is no common version of all binaries found, return None.
+    """
+    if suffixes is None:
+        # No suffixes provided, default to using empty suffix only
+        suffixes = ['']
+
+    for suffix in suffixes:
+        path_map = dict()
+        for name, tool in tools.iteritems():
+            path = which(tool + suffix)
+            if not path:
+                break
+            path_map[name] = path
+        if len(path_map) == len(tools):
+            return Toolchain(**path_map)
+    return None
+
+
+def host_toolchain(xcrun_toolchain='default', tools=None, suffixes=None):
+    """
+    Return a Toolchain with the first available versions of all
+    specified tools, plus clang and clang++, searching in the order of the
+    given suffixes.
+
+    If no matching executables are found, return None.
+    """
+    if tools is None:
+        tools = dict()
+
+    # Require cc and cxx for a toolchain; this ensures whatever llvm tool we
+    # find matches the version of clang returned by a previous run of
+    # this tool
+    tools['cc'] = 'clang'
+    tools['cxx'] = 'clang++'
+
+    if platform.system() == 'Darwin':
+        # Only use xcrun on Darwin
+        path_map = {}
+        for name, tool in tools.iteritems():
+            path = xcrun.find(xcrun_toolchain, tool)
+            if not path:
+                return None
+            path_map[name] = path
+        return Toolchain(**path_map)
+    else:
+        return _first_common_toolchain(tools, suffixes=suffixes)
+
+
+def host_clang(xcrun_toolchain):
+    """
+    Return a Toolchain for the host platform.
+    If no appropriate compilers can be found, return None.
+    """
+    if platform.system() == 'FreeBSD':
+        # See: https://github.com/apple/swift/pull/169
+        # Building Swift from source requires a recent version of the Clang
+        # compiler with C++14 support.
+        freebsd_release_date = _freebsd_release_date()
+        if freebsd_release_date and freebsd_release_date >= 1100000:
+            # On newer releases of FreeBSD, the default Clang is sufficient.
+            return host_toolchain(xcrun_toolchain)
+        else:
+            # On older releases, or on releases for which we cannot determine
+            # the release date, we search for the most modern version
+            # available.
+            return host_toolchain(xcrun_toolchain,
+                                  suffixes=['38', '37', '36', '35'])
+    return host_toolchain(xcrun_toolchain,
+                          suffixes=['-3.8', '-3.7', '-3.6', '-3.5'])
```

Comment:
`suffixes` should contain `'', ` [as before](https://github.com/apple/swift/pull/1586/files#diff-0084283925791f53be5c74fd2938d392L90)